File: packages/sdk-core/src/circuit-breaker.ts
/**
 * Circuit Breaker States
 */
export enum CircuitState {
  /** Normal operation, requests pass through */
  CLOSED = "closed",
  /** Circuit is open, requests fail fast */
  OPEN = "open",
  /** Testing if service has recovered */
  HALF_OPEN = "half_open",
}

/**
 * Circuit Breaker Configuration
 */
export interface CircuitBreakerConfig {
  /** Number of failures before opening circuit (default: 5) */
  failureThreshold: number;
  /** Time to wait before attempting recovery (default: 30000ms) */
  recoveryTimeout: number;
  /** Window for counting failures (default: 60000ms) */
  monitoringWindow: number;
  /** Number of successful requests in half-open to close circuit (default: 3) */
  successThreshold: number;
}

export const DEFAULT_CIRCUIT_BREAKER_CONFIG: CircuitBreakerConfig = {
  failureThreshold: 5,
  recoveryTimeout: 30000,
  monitoringWindow: 60000,
  successThreshold: 3,
};

/**
 * Error thrown when circuit is open
 */
export class CircuitOpenError extends Error {
  constructor(message: string = "Circuit breaker is open") {
    super(message);
    this.name = "CircuitOpenError";
  }
}

/** Event data for circuit breaker events */
interface CircuitEventData {
  state?: CircuitState;
  previousState?: CircuitState;
  failures?: number;
  timeUntilRetry?: number;
  from?: CircuitState;
  to?: CircuitState;
  lastFailureTime?: number;
  successes?: number;
}

type CircuitEventHandler = (data?: CircuitEventData) => void;

/**
 * Circuit Breaker implementation
 *
 * Prevents cascading failures by failing fast when a service is down.
 */
export class CircuitBreaker {
  private state: CircuitState = CircuitState.CLOSED;
  private failures: number[] = [];
  private lastFailureTime: number = 0;
  private halfOpenSuccesses: number = 0;
  private config: CircuitBreakerConfig;
  private eventHandlers: Map<string, CircuitEventHandler[]> = new Map();

  constructor(config: Partial<CircuitBreakerConfig> = {}) {
    this.config = { ...DEFAULT_CIRCUIT_BREAKER_CONFIG, ...config };
  }

  /**
   * Add event listener
   */
  on(event: string, handler: CircuitEventHandler): void {
    const handlers = this.eventHandlers.get(event) || [];
    handlers.push(handler);
    this.eventHandlers.set(event, handlers);
  }

  /**
   * Emit event
   */
  private emit(event: string, data?: CircuitEventData): void {
    const handlers = this.eventHandlers.get(event) || [];
    handlers.forEach((handler) => handler(data));
  }

  /**
   * Execute a function through the circuit breaker
   */
  async execute<T>(fn: () => Promise<T>): Promise<T> {
    // Check if circuit should transition from OPEN to HALF_OPEN
    if (this.state === CircuitState.OPEN) {
      if (this.shouldAttemptReset()) {
        this.transitionTo(CircuitState.HALF_OPEN);
      } else {
        throw new CircuitOpenError(
          `Circuit breaker is open. Will retry after ${this.getTimeUntilRetry()}ms`,
        );
      }
    }

    try {
      const result = await fn();
      this.onSuccess();
      return result;
    } catch (error) {
      this.onFailure();
      throw error;
    }
  }

  /**
   * Handle successful request
   */
  private onSuccess(): void {
    if (this.state === CircuitState.HALF_OPEN) {
      this.halfOpenSuccesses++;
      this.emit("half-open-success", { successes: this.halfOpenSuccesses });

      if (this.halfOpenSuccesses >= this.config.successThreshold) {
        this.reset();
      }
    }

    // Clean up old failures outside monitoring window
    this.cleanupOldFailures();
  }

  /**
   * Handle failed request
   */
  private onFailure(): void {
    const now = Date.now();
    this.failures.push(now);
    this.lastFailureTime = now;

    // If in HALF_OPEN, immediately open the circuit
    if (this.state === CircuitState.HALF_OPEN) {
      this.transitionTo(CircuitState.OPEN);
      this.halfOpenSuccesses = 0;
      return;
    }

    // Clean up old failures and check threshold
    this.cleanupOldFailures();

    if (this.failures.length >= this.config.failureThreshold) {
      this.transitionTo(CircuitState.OPEN);
    }
  }

  /**
   * Remove failures outside the monitoring window
   */
  private cleanupOldFailures(): void {
    const cutoff = Date.now() - this.config.monitoringWindow;
    this.failures = this.failures.filter((t) => t > cutoff);
  }

  /**
   * Check if enough time has passed to attempt reset
   */
  private shouldAttemptReset(): boolean {
    return Date.now() - this.lastFailureTime >= this.config.recoveryTimeout;
  }

  /**
   * Get time until next retry attempt is allowed
   */
  private getTimeUntilRetry(): number {
    const elapsed = Date.now() - this.lastFailureTime;
    return Math.max(0, this.config.recoveryTimeout - elapsed);
  }

  /**
   * Transition to a new state
   */
  private transitionTo(newState: CircuitState): void {
    const oldState = this.state;
    this.state = newState;

    this.emit("state-change", { from: oldState, to: newState });

    if (newState === CircuitState.OPEN) {
      this.emit("circuit-open", {
        failures: this.failures.length,
        lastFailureTime: this.lastFailureTime,
      });
    } else if (newState === CircuitState.CLOSED) {
      this.emit("circuit-closed");
    } else if (newState === CircuitState.HALF_OPEN) {
      this.emit("circuit-half-open");
    }
  }

  /**
   * Reset the circuit breaker to closed state
   */
  private reset(): void {
    this.failures = [];
    this.halfOpenSuccesses = 0;
    this.transitionTo(CircuitState.CLOSED);
  }

  /**
   * Force reset the circuit breaker
   */
  forceReset(): void {
    this.reset();
  }

  /**
   * Force open the circuit breaker
   */
  forceOpen(): void {
    this.lastFailureTime = Date.now();
    this.transitionTo(CircuitState.OPEN);
  }

  /**
   * Get current circuit state
   */
  getState(): CircuitState {
    return this.state;
  }

  /**
   * Get circuit breaker stats
   */
  getStats(): {
    state: CircuitState;
    failures: number;
    lastFailureTime: number | null;
    halfOpenSuccesses: number;
  } {
    return {
      state: this.state,
      failures: this.failures.length,
      lastFailureTime: this.lastFailureTime || null,
      halfOpenSuccesses: this.halfOpenSuccesses,
    };
  }

  /**
   * Check if circuit is allowing requests
   */
  isAllowingRequests(): boolean {
    if (this.state === CircuitState.CLOSED) return true;
    if (this.state === CircuitState.HALF_OPEN) return true;
    if (this.state === CircuitState.OPEN && this.shouldAttemptReset())
      return true;
    return false;
  }
}

File: packages/sdk-python/rollgate/circuit_breaker.py
"""
Circuit Breaker implementation.

Prevents cascading failures by failing fast when a service is down.
"""

import time
from dataclasses import dataclass
from enum import Enum
from typing import Callable, TypeVar, Optional, Awaitable, List

T = TypeVar("T")


class CircuitState(str, Enum):
    """Circuit breaker states."""

    CLOSED = "closed"
    """Normal operation, requests pass through."""

    OPEN = "open"
    """Circuit is open, requests fail fast."""

    HALF_OPEN = "half_open"
    """Testing if service has recovered."""


@dataclass
class CircuitBreakerConfig:
    """Configuration for circuit breaker behavior."""

    failure_threshold: int = 5
    """Number of failures before opening circuit."""

    recovery_timeout_ms: int = 30000
    """Time to wait before attempting recovery."""

    monitoring_window_ms: int = 60000
    """Window for counting failures."""

    success_threshold: int = 3
    """Number of successful requests in half-open to close circuit."""


@dataclass
class CircuitBreakerStats:
    """Statistics about circuit breaker state."""

    state: CircuitState
    failures: int
    last_failure_time: Optional[float]
    half_open_successes: int


class CircuitOpenError(Exception):
    """Raised when circuit breaker is open."""

    def __init__(self, message: str = "Circuit breaker is open", retry_after_ms: int = 0):
        super().__init__(message)
        self.retry_after_ms = retry_after_ms


DEFAULT_CIRCUIT_BREAKER_CONFIG = CircuitBreakerConfig()


class CircuitBreaker:
    """
    Circuit Breaker implementation.

    States:
    - CLOSED: Normal operation, all requests pass through
    - OPEN: Service is down, all requests fail immediately
    - HALF_OPEN: Testing recovery, limited requests allowed
    """

    def __init__(self, config: Optional[CircuitBreakerConfig] = None):
        self._config = config or DEFAULT_CIRCUIT_BREAKER_CONFIG
        self._state = CircuitState.CLOSED
        self._failures: List[float] = []
        self._last_failure_time: float = 0
        self._half_open_successes: int = 0
        self._callbacks: dict[str, list[Callable]] = {
            "state_change": [],
            "circuit_open": [],
            "circuit_closed": [],
            "circuit_half_open": [],
        }

    def on(self, event: str, callback: Callable) -> None:
        """Register an event callback."""
        if event in self._callbacks:
            self._callbacks[event].append(callback)

    def off(self, event: str, callback: Callable) -> None:
        """Remove an event callback."""
        if event in self._callbacks and callback in self._callbacks[event]:
            self._callbacks[event].remove(callback)

    def _emit(self, event: str, *args) -> None:
        """Emit an event to all registered callbacks."""
        for callback in self._callbacks.get(event, []):
            try:
                callback(*args)
            except Exception:
                pass  # Don't let callback errors affect circuit breaker

    async def execute(self, fn: Callable[[], Awaitable[T]]) -> T:
        """
        Execute a function through the circuit breaker.

        Args:
            fn: Async function to execute

        Returns:
            Result of the function

        Raises:
            CircuitOpenError: If circuit is open
            Exception: Any exception from the function
        """
        # Check if circuit should transition from OPEN to HALF_OPEN
        if self._state == CircuitState.OPEN:
            if self._should_attempt_reset():
                self._transition_to(CircuitState.HALF_OPEN)
            else:
                retry_after = self._get_time_until_retry()
                raise CircuitOpenError(
                    f"Circuit breaker is open. Will retry after {retry_after}ms",
                    retry_after_ms=retry_after,
                )

        try:
            result = await fn()
            self._on_success()
            return result
        except Exception as error:
            self._on_failure()
            raise

    def _on_success(self) -> None:
        """Handle successful request."""
        if self._state == CircuitState.HALF_OPEN:
            self._half_open_successes += 1

            if self._half_open_successes >= self._config.success_threshold:
                self._reset()

        # Clean up old failures outside monitoring window
        self._cleanup_old_failures()

    def _on_failure(self) -> None:
        """Handle failed request."""
        now = time.time() * 1000  # Convert to ms
        self._failures.append(now)
        self._last_failure_time = now

        # If in HALF_OPEN, immediately open the circuit
        if self._state == CircuitState.HALF_OPEN:
            self._transition_to(CircuitState.OPEN)
            self._half_open_successes = 0
            return

        # Clean up old failures and check threshold
        self._cleanup_old_failures()

        if len(self._failures) >= self._config.failure_threshold:
            self._transition_to(CircuitState.OPEN)

    def _cleanup_old_failures(self) -> None:
        """Remove failures outside the monitoring window."""
        cutoff = time.time() * 1000 - self._config.monitoring_window_ms
        self._failures = [t for t in self._failures if t > cutoff]

    def _should_attempt_reset(self) -> bool:
        """Check if enough time has passed to attempt reset."""
        elapsed = time.time() * 1000 - self._last_failure_time
        return elapsed >= self._config.recovery_timeout_ms

    def _get_time_until_retry(self) -> int:
        """Get time until next retry attempt is allowed."""
        elapsed = time.time() * 1000 - self._last_failure_time
        return max(0, int(self._config.recovery_timeout_ms - elapsed))

    def _transition_to(self, new_state: CircuitState) -> None:
        """Transition to a new state."""
        old_state = self._state
        self._state = new_state

        self._emit("state_change", old_state, new_state)

        if new_state == CircuitState.OPEN:
            self._emit("circuit_open", len(self._failures), self._last_failure_time)
        elif new_state == CircuitState.CLOSED:
            self._emit("circuit_closed")
        elif new_state == CircuitState.HALF_OPEN:
            self._emit("circuit_half_open")

    def _reset(self) -> None:
        """Reset the circuit breaker to closed state."""
        self._failures = []
        self._half_open_successes = 0
        self._transition_to(CircuitState.CLOSED)

    def force_reset(self) -> None:
        """Force reset the circuit breaker (for testing/manual recovery)."""
        self._reset()

    def force_open(self) -> None:
        """Force open the circuit breaker (for testing/manual circuit trip)."""
        self._last_failure_time = time.time() * 1000
        self._transition_to(CircuitState.OPEN)

    @property
    def state(self) -> CircuitState:
        """Get current circuit state."""
        return self._state

    def get_stats(self) -> CircuitBreakerStats:
        """Get circuit breaker statistics."""
        return CircuitBreakerStats(
            state=self._state,
            failures=len(self._failures),
            last_failure_time=self._last_failure_time if self._last_failure_time else None,
            half_open_successes=self._half_open_successes,
        )

    def is_allowing_requests(self) -> bool:
        """Check if circuit is allowing requests."""
        if self._state == CircuitState.CLOSED:
            return True
        if self._state == CircuitState.HALF_OPEN:
            return True
        if self._state == CircuitState.OPEN and self._should_attempt_reset():
            return True
        return False
